lib/api.js

```diff
@@ -0,0 +1,213 @@
+'use strict';
+var EventEmitter = require('events').EventEmitter;
+var path = require('path');
+var util = require('util');
+var fs = require('fs');
+var flatten = require('arr-flatten');
+var Promise = require('bluebird');
+var figures = require('figures');
+var assign = require('object-assign');
+var globby = require('globby');
+var chalk = require('chalk');
+var fork = require('./fork');
+
+function Api(options) {
+	if (!(this instanceof Api)) {
+		return new Api(options);
+	}
+
+	EventEmitter.call(this);
+
+	assign(this, options);
+
+	this.rejectionCount = 0;
+	this.exceptionCount = 0;
+	this.passCount = 0;
+	this.failCount = 0;
+	this.fileCount = 0;
+	this.testCount = 0;
+	this.errors = [];
+	this.stats = [];
+	this.tests = [];
+
+	Object.keys(Api.prototype).forEach(function (key) {
+		this[key] = this[key].bind(this);
+	}, this);
+}
+
+util.inherits(Api, EventEmitter);
+module.exports = Api;
+
+Api.prototype._runFile = function (file) {
+	var args = [file];
+
+	if (this.failFast) {
+		args.push('--fail-fast');
+	}
+
+	if (this.serial) {
+		args.push('--serial');
+	}
+
+	// Forward the `time-require` `--sorted` flag.
+	// Intended for internal optimization tests only.
+	if (this._sorted) {
+		args.push('--sorted');
+	}
+
+	return fork(args)
+		.on('stats', this._handleStats)
+		.on('test', this._handleTest)
+		.on('unhandledRejections', this._handleRejections)
+		.on('uncaughtException', this._handleExceptions);
+};
+
+Api.prototype._handleRejections = function (data) {
+	this.emit('rejection', data);
+	this.rejectionCount += data.rejections.length;
+};
+
+Api.prototype._handleExceptions = function (data) {
+	this.emit('exception', data);
+	this.exceptionCount++;
+};
+
+Api.prototype._handleStats = function (stats) {
+	this.testCount += stats.testCount;
+};
+
+Api.prototype._handleTest = function (test) {
+	test.title = this._prefixTitle(test.file) + test.title;
+
+	var isError = test.error.message;
+
+	if (isError) {
+		this.errors.push(test);
+	} else {
+		test.error = null;
+	}
+
+	this.emit('test', test);
+};
+
+Api.prototype._prefixTitle = function (file) {
```

Comment:
@jamestalmage We already do this in https://github.com/sindresorhus/ava/blob/e2e45151dc5ac249d49a7cd94baca8c70f9ed2ce/cli.js#L57